flink-connector/docs/content/docs/connectors/datastream/pubsub.md

```diff
@@ -0,0 +1,158 @@
+# Google Cloud Pub/Sub Connector
+
+This connector provides access to reading data from and writing data to 
+[Google Cloud Pub/Sub](https://cloud.google.com/pubsub).
+
+## Pub/Sub Source
+
+Pub/Sub source streams data from a single Google Cloud Pub/Sub subscription. The
+sample below shows the minimal configurations required to build Pub/Sub source.
+
+```java
+PubSubSource.<String>builder()
+     .setDeserializationSchema(PubSubDeserializationSchema.dataOnly(new SimpleStringSchema()))
+     .setProjectName("my-project-name")
+     .setSubscriptionName("my-subscription-name")
+    .build()
+```
+
+### Subscription
+
+Pub/Sub source only supports streaming messages from
+[pull subscriptions](https://cloud.google.com/pubsub/docs/pull). Push
+subscriptions are not supported.
+
+Pub/Sub source can create parallel readers to to the same subscription. Since
+Google Cloud Pub/Sub has no notion of subscription partitions or splits, a
+message can be received by any reader. Google Cloud Pub/Sub automatically load
+balances message delivery across readers.
+
+### Deserialization Schema
+
+`PubSubDeserializationSchema<T>` is required to define how
+[`PubsubMessage`](https://cloud.google.com/pubsub/docs/reference/rest/v1/PubsubMessage)
+is deserialized into an output that is suitable for processing.
+
+For convenience, `PubSubDeserializationSchema.dataOnly` can be used if only the
+field `PubsubMessage.data` is required for processing.
+
+### Boundedness
+
+Pub/Sub source streams unbounded data, only stopping when a Flink job stops or
+fails.
+
+### Checkpointing
+
+Checkpointing is required to use Pub/Sub source. When a checkpoint completes,
+all messages delivered before the last successful checkpoint are acknowledged to
+Google Cloud Pub/Sub. In case of failure, messages delivered after the last
+successful checkpoint are unacknowledged and will automatically be redelivered.
+Note that there is no message delivery state stored in checkpoints, so retained
+checkpoints are not necessary to resume using Pub/Sub source.
+
+### Flow Control
+
+Each Pub/Sub source subtask manages a `StreamingPull` connection to Google Cloud
+Pub/Sub. That is, a Pub/Sub source with parallelism of 20 manages 20 separate
+`StreamingPull` connections to Google Cloud Pub/Sub. The flow control settings
+described in this section are applied to **individual** connections.
+
+Several Pub/Sub source options configure flow control. These options are
+illustrated below:
+
+```java
+PubSubSource.<OutputT>builder()
+    // Allow up to 10,000 message deliveries per checkpoint interval.
+    .setMaxOutstandingMessagesCount(10_000L)
```

Comment:
do newer java versions still support underscores in numeric literals? (genuine question; i don't know the answer)

Reply:
I can't find any evidence of support being removed. This should be okay as long as the Java version is 7 or later.